core/src/main/java/org/apache/calcite/rel/rules/DpHyp.java

```diff
@@ -173,42 +193,97 @@ private void enumerateCmpRec(long csg, long cmp, long forbidden) {
   private void emitCsgCmp(long csg, long cmp, List<HyperEdge> edges) {
     RelNode child1 = dpTable.get(csg);
     RelNode child2 = dpTable.get(cmp);
-    if (child1 == null || child2 == null) {
-      throw new IllegalArgumentException(
-          "csg and cmp were not enumerated in the previous dp process");
-    }
+    ImmutableList<HyperGraph.NodeState> csgOrder = resultInputOrder.get(csg);
+    ImmutableList<HyperGraph.NodeState> cmpOrder = resultInputOrder.get(cmp);
+    assert child1 != null && child2 != null && csgOrder != null && cmpOrder != null;
+    assert Long.bitCount(csg) == csgOrder.size() && Long.bitCount(cmp) == cmpOrder.size();
 
     JoinRelType joinType = hyperGraph.extractJoinType(edges);
     if (joinType == null) {
       return;
     }
-    RexNode joinCond1 = hyperGraph.extractJoinCond(child1, child2, edges);
+    // verify whether the subgraph is legal by using the conflict rules in hyperedges
+    if (!hyperGraph.applicable(csg | cmp, edges)) {
+      return;
+    }
+
+    List<HyperGraph.NodeState> unionOrder = new ArrayList<>(csgOrder);
+    unionOrder.addAll(cmpOrder);
+    // build join condition from hyperedges. e.g.
```

Comment:
Could you explain what you are going to do with a sql or plan? These are a bit difficult to understand for people who don't know much about the paper, and the same for other places where formulas are used as comments.